include/mlir/IR/OpImplementation.h

```diff
@@ -410,10 +410,25 @@ class OpAsmParser {
                                           ArrayRef<OperandType> arguments,
                                           ArrayRef<Type> argTypes) = 0;
 
-  /// Parse a region argument.  Region arguments define new values, so this also
-  /// checks if the values with the same name has not been defined yet.
+  /// Parse a region argument.  Region arguments define new values; so this also
+  /// checks if values with the same name have not been defined yet.
   virtual ParseResult parseRegionArgument(OperandType &argument) = 0;
 
+  /// Parse zero or more region arguments with a specified surrounding
+  /// delimiter, and an optional required argument count. Region arguments
+  /// define new values; so this also checks if values with the same names have
+  /// not been defined yet.
+  virtual ParseResult
+  parseRegionArgumentList(SmallVectorImpl<OperandType> &result,
+                          int requiredOperandCount = -1,
```

Comment:
Is there a test case for Delimiter::None?

Reply:
Actually, there isn't. The only way I can think of testing it is by introducing an op that has such a custom form. AFAIK, the gpu.launch is the only op in MLIR that has multiple region arguments; I have another experimental one on the side, but that isn't meant for mainstream yet and it too uses a parenthesis enclosed list.

Reply:
You should be able to test by adding a fake operation to the TestDialect. I added that specifically for the things that don't have a use case in-tree.

https://github.com/tensorflow/mlir/blob/master/test/lib/TestDialect/TestOps.td

Reply:
Just looked at TestOps.td:

Reply:
Just looked at TestOps.td. Does "Arguments< ..>" there mean operands (cf. appended lines for example)? An operand is expected to have a def prior to its use, while in contrast a region argument is expected not to have a def at all. If "Arguments" in the .td means operands, what are the region arguments referred to? It looks like the naming is confusing here?

> def OpD : TEST_Op<"op_d">, Arguments<(ins I32:$arg)>, Results<(outs I32:$res)>;
> def OpE : TEST_Op<"op_e">, Arguments<(ins I32:$arg)>, Results<(outs I32:$res)>;
> def OpF : TEST_Op<"op_f">, Arguments<(ins I32:$arg)>, Results<(outs I32:$res)>;
> def OpG : TEST_Op<"op_g">, Arguments<(ins I32:$arg)>, Results<(outs I32:$res)>;

Reply:
> You should be able to test by adding a fake operation to the TestDialect. I added that specifically for the things that don't have a use case in-tree.
> 
> https://github.com/tensorflow/mlir/blob/master/test/lib/TestDialect/TestOps.td

Is there an example of a test op that has a parse method for its custom form? I couldn't immediately find any. (Just adding a test op that takes multiple region arguments isn't sufficient here.)

Reply:
> Just looked at TestOps.td. Does "Arguments< ..>" there mean operands (cf. appended lines for example)? An operand is expected to have a def prior to its use, while in contrast a region argument is expected not to have a def at all. If "Arguments" in the .td means operands, what are the region arguments referred to? It looks like the naming is confusing here?
> 

Arguments here refer to operands and attributes (consult the docs directory for ODS).

> Is there an example of a test op that has a parse method for its custom form? I couldn't immediately find any. (Just adding a test op that takes multiple region arguments isn't sufficient here.)

There isn't feel free to add one that exercises this path.

Reply:
In the ODS, I see "operands and attributes" but nothing on regions and thus region arguments. So any op that has a region can't be tablegen'd currently(?), and this test op will have to be done via TestDialect.h and TestDialect.cpp?

Reply:
There are regions there, but the region arguments one seems to be pending still, so via C++ SGTM.

Reply:
Done now (via TestDialect.cpp).